0023-communication-protocol.adoc

```diff
@@ -0,0 +1,383 @@
+
+== Recommendation
+
+Based on the research above, this section presents a proposal for a set of libp2p protocols that can be combined to offer the required functionality.
+
+=== Announcing a swap
+
+Protocol-identifier: `/comit/swap/announce/1.0.0`
+
+This protocol includes two messages:
+
+. The announce message
+. The confirmation message
+
+==== Announce-message
+
+[source,json]
+----
+{
+  "swap_digest": "12202c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae"
+}
+----
+
+This message is always sent by Alice.
+See <<Assumptions and goals,assumptions point 3>> for further explanation.
+
+The `swap_digest` field is a https://github.com/multiformats/multihash[multihash-encoded] root hash of a hash list that contains all details of the swap.
+It remains to be defined, how the digest is exactly computed but the idea is the following:
+
+. Take all of the swaps information _known to the COMIT node_ (alpha ledger, alpha asset, protocols, etc) and arrange it in a list
+. Serialize each item into a yet-to-be defined representation
+. Hash each list entry
+. Sort the hashes lexically
+. Compute the root hash
+
+The idea is that, independently of what information about the swap is given to the COMIT node (imagine different usecases, with different negotiation protocols on top of COMIT), there is a deterministic way of computing this digest.
+If both parties, independently of each other, compute the same digest, they must be knowledgeable of the same parameters.
+This should give us enough certainty to start executing the swap assuming the other party is in good will.footnote:[It is pretty much impossible to fully protected against bad will upfront.]
+Only sending a digest has several advantages over sending the actual data of the swap for comparison:
+
+- The message format is much simpler and can be reused for every swap-protocol, independent of the cryptographic protocol in use.
+- The message format is agnostic over what the nodes are actually committing to.
+
+The nodes know, which data they used to compute the digest.
+Consequently, they are also aware of what is still missing to start the swap.
+Imagine a negotiation protocol that doesn't include expiriesfootnote:[Despite the fact that they are price-relevant.].
+The nodes could:
+
+.. either default to certain expiries and include them in the digest
+.. or simply not include them in the digest.
+
+In case of a), a matching digest means both nodes used the same default expiries.
+In case of b), a matching digest means both nodes are aware that the expiries are still missing and they will have to somehow negotiate them.
+
+==== Confirmation message
+
+[source,json]
+----
+{
+  "swap_id": "713b0c93-53a1-46b2-9418-de8d46750b79"
+}
+----
+
+In case Bob knows about a swap with the same digest, he generates a UUIDv4 and sends it back to Alice.
+If he doesn't know about a swap with this digest, he closes the substream without sending any message.
```

Comment:
Let my rephrase to see if I understand the process correctly: 

* Alice adds to her cnd some swap information
* Bob does the same thing
** both generate the `swap_digest`
* Alice sends to Bob this `swap_digest`
* If Bob knows the `swap_digest` he generates a UUIDv4 (point 3.1.2 Confirmation message) and returns it to Alice.
* ...

Isn't there a danger of a race condition where Alice sends something to Bob but Bob does not expect anything from Alice yet?

Reply:
Your understanding is correct.

> Isn't there a danger of a race condition where Alice sends something to Bob but Bob does not expect anything from Alice yet?

Yes that exists but I am not sure it can be solved completely.
Bob can wait for some time before rejecting the swap but it shouldn't be too long.
Generally, whatever lead the application to instruct cnd to do a swap, the same thing should happen on the other side.

Reply:
Should Bob in this case indicate that he doesn't know what Alice is talking about? Meaning, return something?

Reply:
For example?

A string that Alice can print to the logs?
Should she be able to automatically recover from that?

That topic is a whole new can of worms.
My personal preference would be to evaluate that the need for this as we go and use it ourselves and have other people use it.
That should tell us about common error scenarios and we can think about how to attack them.

I see the general problem, but I don't have a feasible solution at this stage.